include/dos_system.h

```diff
@@ -30,6 +30,7 @@
 
 #define DOS_NAMELENGTH 12
 #define DOS_NAMELENGTH_ASCII (DOS_NAMELENGTH+1)
+#define LFN_NAMELENGTH 255
```

Comment:
For FAT drives the Unicode characters will be converted to ASCII characters by the later code not yet in this Part 1. What is your suggestion?

Reply:
Sounds like we need to retain the unicode characters as is.

Otherwise, we would read the unicode filename from host OS, convert unicode name to ascii  pass it into the DOS 7.0 application, and the user could no longer save edits to the file back to the original, because the conversion is "one way".  So they could only save a new ascii-only filename out to the host OS.

Reply:
@kcgen Even though on FAT drives the LFN names are saved on Unicode format, the DOS LFN APIs themselves use ASCII characters to communicate, after Unicode to local codepage conversions, and vice versa. Since DOS programs make use of LFN support via the LFN APIs, basically the LFN implementation for drives will take care of the Unicode => codepage and codepage => Unicode conversions, not by the DOS programs themselves (DOS programs will work on the local codepages instead).

Reply:
Oh, very it interesting @Wengier; thanks for explaining.

Just for my own knowledge, so the codepage retails enough info to go back and forth without loss to and from the same underlying unicode filename?  And the ascii conversion really is just superficial into the dos program via the API? (But ultimately still retains the codepage name underneath the API?).

Reply:
@kcgen They are two types of code pages, most are single byte (SBCS) code pages, and the rest are double byte (DBCS) code pages, as used by CJK (Chinese, Japanese, Korean) characters. DBCS codepages are much bigger than SBCS codepages in order to contain most (if not all) used characters in the specific languages. DOS programs generally work with a specific codepage (or language) instead of Unicode, so there is typically no problem in such conversions as long as that code page is used during both conversions (back and forth). On the other hand, if there is a code page mismatch, then an erroneous conversion is likely expected. If you look at the DOSLFN software (LFN driver for DOS), you will find a bunch of codepages (.TBL files) included in the package for Unicode conversion, and you are expected to use the correct code page for your language (and DOS software), otherwise you may see either incorrectly converted characters or underlines for fail-converted characters. So again DOS programs will simply work on local codepages, and all Unicode conversions will be handled by the LFN implementation instead.

Reply:
Thanks; this is starting to make sense.
So the variable `LFN_NAMELENGTH`, is not the number of bytes in the unicode filename (which presumably would need up to 2x or maybe 4x the bytes?), but is the the converted local codepage length?

And the conversion tables: do those always convert from some expanded unicode values down to single char/ascii bytes? So we're guaranteed that all the conversion tables always spit out a max up 255 chars, no matter how bloated the unicode name is in the underlying fat32 filesystem?

Reply:
@kcgen Yes, ``LFN_NAMELENGTH`` is the codepage-converted ASCII length. The Unicode conversion tables will convert the Unicode values to local codepages if possible, otherwise it will be converted to e.g. underscores instead. But in the end the LFN name will be 255 ASCII chars maximum.

Reply:
@dreamer As stated earlier, the Unicode conversion table is included in the LFN implementation for conversions between Unicode and ASCII characters. Please don't expect DOS programs to handle Unicode characters - as a general rule DOS programs only deal with ASCII characters, so it is the job of the LFN code to take care of the Unicode conversion part, and LFN_NAMELENGTH is simply **not** intended for Unicode characters. If there is an error occurred during this conversion, and an error flag will be returned to the DOS program by the LFN API to let the DOS programs know that the Unicode conversions may not be reliable and to be treated with caution. This is how LFN is handled, and DOS programs will **never** deal with the Unicode directly.

Reply:
@dreamer You can also take a look at the LFN API yourself, e.g. "LONG FILENAME - CHANGE DIRECTORY":

http://www.ctyme.com/intr/rb-3199.htm

The parameter is expected to be a ASCIZ long name of directory, not a Unicode name at all. ASCIZ simply means ASCII followed by a zero byte, nothing to do with Unicode.